### PR TITLE
[github] Add labeler action

### LINKED ISSUE #126002: Migrate cocoon label code to github workflows
There's code like https://cs.opensource.google/flutter/cocoon/+/main:app_dart/lib/src/request_handlers/github/webhook_subscription.dart;l=352 that can be moved into the repo under test.

This should make it easier to find, and quicker to deploy.

* [x] Add docs on adding new labels - https://github.com/flutter/flutter/wiki/Labeling-PRs
* [x] Migrate flutter/flutter - https://github.com/flutter/flutter/pull/126012
* [x] Migrate flutter/engine - https://github.com/flutter/engine/pull/42566
* [x] Migrate flutter/packages - https://github.com/flutter/packages/pull/335
* [x] Remove label functionality from Cocoon - https://github.com/flutter/cocoon/pull/2687

### PR DESCRIPTION
**Pull request by CaseyHillers**
Part of https://github.com/flutter/flutter/issues/126002

Migrate the [Cocoon logic for labelling](https://cs.opensource.google/flutter/cocoon/+/main:app_dart/lib/src/request_handlers/github/webhook_subscription.dart;l=352) directly into the repo under test

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

### CODE REVIEW

File: .github/labeler.yml [hunk 1]
```diff
@@ -0,0 +1,109 @@
+'a: accessibility':
+  - any:
+    - accessibility
+    - semantics
```
Comment (keyonghan) [inline .github/labeler.yml:4]: For all `pathContainsLabels`, these should be `**...**`.

File: .github/labeler.yml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+framework:
+  - any:
+    - packages/flutter/**
+    - packages/flutter_driver/**
+    - packages/flutter_goldens/**
+    - packages/flutter_goldens_client/**
+    - packages/flutter_test/**
+    - packages/integration_test/**
+
+integration-test:
```
Comment (keyonghan) [inline .github/labeler.yml:77]:
```suggestion
integration_test:
```

File: .github/labeler.yml [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    - dev/**
+    - examples/**
+    - packages/flutter_goldens/**
+    - packages/flutter_goldens_client/**
+
+'team: gallery':
+  - any:
+    - examples/flutter_gallery/**
+
+tech_debt:
```
Comment (keyonghan) [inline .github/labeler.yml:100]:
```suggestion
tech-debt:
```

**Review by keyonghan — COMMENTED** [review]
How do we validate the actions work as expected

File: .github/labeler.yml [hunk 4]
```diff
@@ -0,0 +1,109 @@
+'a: accessibility':
```
Comment (godofredoc) [inline .github/labeler.yml:1]: Docs?

File: .github/workflows/labeler.yml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+- pull_request_target
+
+jobs:
+  triage:
+    permissions:
+      contents: read
+      pull-requests: write
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/labeler@v4
```
Comment (godofredoc) [inline .github/workflows/labeler.yml:12]: pin to hash, dependabot will update it automatically.

File: .github/workflows/labeler.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+name: "Pull Request Labeler"
+on:
+- pull_request_target
+
```
Comment (godofredoc) [inline .github/workflows/labeler.yml:4]: please set read-all as default.

File: .github/workflows/labeler.yml [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+jobs:
+  triage:
+    permissions:
+      contents: read
+      pull-requests: write
```
Comment (godofredoc) [inline .github/workflows/labeler.yml:9]: Is there any more granular permission that can be used here?

File: .github/workflows/labeler.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+name: "Pull Request Labeler"
```
Comment (godofredoc) [inline .github/workflows/labeler.yml:1]: Docs? license header?

File: .github/workflows/labeler.yml (hunk 7 above)
Comment (CaseyHillers) [reply .github/workflows/labeler.yml:16]: Unfortunately, it seems the pull request permission is a catch all. Autosubmit seems to have this same issue where it requires write access to edit the labels on PRs.

I filed a support ticket with GitHub at https://support.github.com/ticket/enterprise/1732/2139514 to see if they can allow smaller ACLs on pull request.

**Comment by CaseyHillers** [comment]
> How do we validate the actions work as expected

In the interim, Cocoon and the action can both run. If we notice discrepancies, we can update the workflow here. After ~100 PRs, we should be able to remove the logic from Cocoon.

File: .github/labeler.yml (hunk 1 above)
Comment (keyonghan) [reply .github/labeler.yml:9]: > > How do we validate the actions work as expected
> 
> In the interim, Cocoon and the action can both run. If we notice discrepancies, we can update the workflow here. After ~100 PRs, we should be able to remove the logic from Cocoon.

How do we detect discrepancies between these two logics?

**Comment by keyonghan** [comment]
> > How do we validate the actions work as expected
> 
> In the interim, Cocoon and the action can both run. If we notice discrepancies, we can update the workflow here. After ~100 PRs, we should be able to remove the logic from Cocoon.

How do we detect discrepancies between these two logics?

**Comment by CaseyHillers** [comment]
> How do we detect discrepancies between these two logics?

I plan to manually triage open PRs for the week to make sure the GitHub action is setting all the labels, and not flutter-dashboard. I expect the GitHub action will get priority as it's running directly on GitHub, and has no other work to do (like the webhook subscription does)

**Comment by keyonghan** [comment]
> > How do we detect discrepancies between these two logics?
> 
> I plan to manually triage open PRs for the week to make sure the GitHub action is setting all the labels, and not flutter-dashboard. I expect the GitHub action will get priority as it's running directly on GitHub, and has no other work to do (like the webhook subscription does)

I am wondering if any logs available for the defined action, but a manual look should tell. LGTM as this is not a blocker even something happens with cocoon around.

**Review by keyonghan — APPROVED** [review]
LGTM

**Comment by CaseyHillers** [comment]
@xilaizhang any idea why this PR isn't getting Google Testing results on it?

**Review by itsjustkevin — APPROVED** [review]

**Comment by XilaiZhang** [comment]
> @XilaiZhang any idea why this PR isn't getting Google Testing results on it?

umm looks like this still has notify github set to false, did we go/upate-frob the legacy job?

**Comment by CaseyHillers** [comment]
Last roll out of those jobs was May 1, did we miss deploying a recent fix?

**Comment by XilaiZhang** [comment]
> Last roll out of those jobs was May 1, did we miss deploying a recent fix?

Right yeah then it wouldn't include cl/529188369 from yesterday which corrects the notify github behavior.

**Comment by XilaiZhang** [comment]
redeployed legacy frob server just now to include yesterday's fix. this was also seen in b/274947171 and impacting other prs. 

(funny how i was joking in triage yesterday this notifyGithub bug would cause a P0 or P1 soon, and it did prove itself today)

**Comment by XilaiZhang** [comment]
This is good to merge (clicking on details page shows 0 errors).

we can either 1. manually set github status to pass, or 2. push a new commit to propagate the testing result.

**Review by godofredoc — APPROVED** [review]
Will there be a follow up to remove this functionality from cocoon and adding documentation on how to add new labels?

**Pull request merged**